### PR TITLE
profile: replace otherData.date field with profile_finish_ts

### PR DESCRIPTION
**Pull request by sluongng**
"date" field in Bazel json profile is a confusing field.
- It uses Java Date.toString() output which includes local timezone that could be hard to parse.

- The name is misleading since users could mistook it for profile start time, while it actually means the finish/end time, when Bazel calls the writer to serialize profiling data to disk.

Let's deprecate the "date" field and remove it from documentation. Replace it with "profile_finish_ts" which has a clearer name and value is more consistent with the time unit being used in the rest of the JSON profile (microseconds).

### CODE REVIEW

**Comment by sluongng** [comment]
@fweikert I think you own the Bazel profiling code, could you please take a look?

**Comment by sgowroji** [comment]
Hi @sluongng, Its already shared with @fweikert for review. I will import once approved. Thanks !

File: src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java [hunk 1]
```diff
@@ -172,13 +172,16 @@ public void run() {
...
                   new BufferedOutputStream(outStream, 262144), StandardCharsets.UTF_8))) {
+        var finishDate = new Date();
         writer.beginObject();
         writer.name("otherData");
         writer.beginObject();
         writer.name("bazel_version").value(BlazeVersionInfo.instance().getReleaseName());
         writer.name("build_id").value(buildID.toString());
         writer.name("output_base").value(outputBase);
-        writer.name("date").value(new Date().toString());
+        // deprecated, use profile_finish_ts instead
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java:182]: Let's not deprecate the field, it is much easier for humans to parse than a timestamp :-)

**Review by meisterT — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java [hunk 2]
```diff
@@ -172,13 +172,16 @@ public void run() {
               // The buffer size of 262144 is chosen at random.
               new OutputStreamWriter(
                   new BufferedOutputStream(outStream, 262144), StandardCharsets.UTF_8))) {
+        var finishDate = new Date();
```
[the remark this answers is not in the document]
Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java]: I like Instant much better. Thanks for this suggestion.

**Comment by sluongng** [comment]
It has been relatively long time so I lost part of the context for this change.

Took me a bit to review this to see why I needed it: For translating Bazel Profile to Open Telemetry data format, I need a parable time stamp.

Updated according to review feedbacks.

File: src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java (hunk 1 above)
Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java]: I don't like how ambiguous the field name is though. It misled me to think it was build start time but in reality, it's build finished time.

Do you think we could rename it?

File: site/en/configure/json-trace-profile.md [hunk 3]
```diff
@@ -167,7 +167,7 @@ Example:
 {
   "otherData": {
     "build_id": "101bff9a-7243-4c1a-8503-9dc6ae4c3b05",
-    "date": "Wed Oct 26 08:22:35 CEST 2022",
```
Comment (meisterT) [inline site/en/configure/json-trace-profile.md:170]: please keep the date in the example

File: src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java (hunk 1 above)
Comment (meisterT) [reply src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java]: We could, but let's do it separately. This change can be cherry-picked then into a minor version and we only change the field name in the next major version.

**Comment by sluongng** [comment]
@meisterT I updated both the diff and the commit message / PR summary accordingly.

Do you want to introduce the deprecation patch or shall I?
I do think removing the ambiguous field will provide a much better UX vs leaving it there.

**Review by meisterT — APPROVED** [review]

**Comment by meisterT** [comment]
> @meisterT I updated both the diff and the commit message / PR summary accordingly.

Thanks!

 
> Do you want to introduce the deprecation patch or shall I? I do think removing the ambiguous field will provide a much better UX vs leaving it there.

Feel free to send a separate PR for this!

**Comment by brentleyjones** [comment]
@bazel-io flag

**Comment by keertk** [comment]
@bazel-io fork 6.2.0

**Comment by keertk** [comment]
Hi @sluongng if we want to include this in 6.2.0, it looks like some other commits need to be cherry-picked first (e.g. https://github.com/bazelbuild/bazel/commit/eb279aafdacb5f66c1c3a4c5aae2746fb93b2e57?). Could you please submit a PR against the `release-6.2.0` branch with the required changes? Thanks!
cc @meisterT

**Comment by meisterT** [comment]
There were many changes we did not cherry pick into 6.1. If we want this functionality in 6.2, I think the easiest is to repeat the spirit of this change instead of cherrypicking other changes.

**Comment by sluongng** [comment]
Recreated in https://github.com/bazelbuild/bazel/pull/18129

**Comment by timothyg-stripe** [comment]
Hi @sluongng, I don't think this change is correct. You wrote that 

> The name is misleading since users could mistook it for profile start time, while it actually means the finish/end time, when Bazel calls the writer to serialize profiling data to disk.

but this method is actually called at the beginning of the Bazel process.

Additionally, `profile_finish_ts` lacks precision compared to `date`. On JDK 17, `Instant.now().toString()` gives us microsecond precision on Linux, while `profile_finish_ts` is rounded to the second.
